hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/handlers/ContractUpdateHandler.java

```diff
@@ -74,6 +107,124 @@ private boolean hasCryptoAdminKey(final ContractUpdateTransactionBody op) {
+        validateTrue(toBeUpdated != null, INVALID_CONTRACT_ID);
+
+        final var changed = update(toBeUpdated, context, op);
+
+        context.serviceApi(TokenServiceApi.class).updateContract(changed);
+    }
+
+    public Account update(
+            @NonNull final Account contract,
+            @NonNull final HandleContext context,
+            @NonNull final ContractUpdateTransactionBody op) {
+
+        validate(contract, context, op);
+
+        final var builder = contract.copyBuilder();
+        if (op.hasAdminKey()) {
+            if (IMMUTABILITY_SENTINEL_KEY.equals(op.adminKey())) {
+                builder.key(contract.key());
+            } else {
+                builder.key(op.adminKey());
+            }
+        }
+        if (op.hasExpirationTime()) {
+            if (contract.expiredAndPendingRemoval()) {
+                builder.expiredAndPendingRemoval(false);
+            }
+            builder.expirationSecond(op.expirationTime().seconds());
+        }
+        if (op.hasAutoRenewPeriod()) {
+            builder.autoRenewSeconds(op.autoRenewPeriod().seconds());
+        }
+        if (affectsMemo(op)) {
+            final var newMemo = op.hasMemoWrapper() ? op.memoWrapper() : op.memo();
+            context.attributeValidator().validateMemo(newMemo);
+            builder.memo(newMemo);
+        }
+        if (op.hasStakedAccountId()) {
+            if (SENTINEL_ACCOUNT_ID.equals(op.stakedAccountId())) {
+                builder.stakedAccountId((AccountID) null);
+            } else {
+                builder.stakedAccountId(op.stakedAccountId());
+            }
+        } else if (op.hasStakedNodeId()) {
+            builder.stakedNodeId(op.stakedNodeId());
+        }
+        if (op.hasDeclineReward()) {
+            builder.declineReward(op.declineReward());
+        }
+        if (op.hasAutoRenewAccountId()) {
+            builder.autoRenewAccountId(op.autoRenewAccountId());
+        }
+        if (op.hasMaxAutomaticTokenAssociations()) {
+            final var ledgerConfig = context.configuration().getConfigData(LedgerConfig.class);
+            if (op.maxAutomaticTokenAssociations() > ledgerConfig.maxAutoAssociations()) {
+                throw new HandleException(REQUESTED_NUM_AUTOMATIC_ASSOCIATIONS_EXCEEDS_ASSOCIATION_LIMIT);
+            }
+            builder.maxAutoAssociations(op.maxAutomaticTokenAssociations());
+        }
+        return builder.build();
+    }
+
+    private void validate(Account contract, HandleContext context, ContractUpdateTransactionBody op) {
+        if (op.hasAdminKey()) {
+            boolean keyNotSentinel = !IMMUTABILITY_SENTINEL_KEY.equals(op.adminKey());
+            boolean keyIsUnset = op.adminKey().key().kind() == KeyOneOfType.UNSET;
+            boolean keyIsNotValid = !KeyUtils.isValid(op.adminKey());
+            if (keyNotSentinel && (keyIsUnset || keyIsNotValid)) {
+                throw new HandleException(INVALID_ADMIN_KEY);
+            }
```

Comment:
you can use `validateFalse(keyNotSentinel && (keyIsUnset || keyIsNotValid), INVALID_ADMIN_KEY`